File: Data Structures & Algorithms/matrixBFS/submission-0.js
class Solution {
    /**
     * @param {number[][]}
     * @returns {number}
     */
    shortestPath(grid) {
        const rows = grid.length, cols = grid[0].length;
        let queue = new Queue();
        let visit = new Set();
        queue.push([0,0]);
        visit.add('0,0');
        let length = 0;
        const directions = [[0,1], [0,-1],[1,0],[-1,0]];

        while(!queue.isEmpty()){
            let size = queue.size();
            console.log(size)
            for (let i = 0; i < size; i++){
                let [r,c] = queue.pop();
                if (r === rows - 1 && c === cols - 1){
                    return length;
                }
                for (let [dr, dc] of directions){
                    let nr = dr + r, nc = dc + c;
                    if (
                        Math.min(nr, nc) < 0 ||
                        nr === rows ||
                        nc === cols ||
                        visit.has(`${nr},${nc}`) ||
                        grid[nr][nc] === 1
                    ){
                        continue;
                    }
                    queue.push([nr,nc]);
                    visit.add(`${nr},${nc}`);
                }
            }

            length += 1;
        }

        return -1;
    }
}

File: Data Structures & Algorithms/islands-and-treasure/submission-0.py
class Solution:
    def islandsAndTreasure(self, grid: List[List[int]]) -> None:
        ROWS, COLS = len(grid), len(grid[0])
        q = collections.deque()
        for r in range(ROWS):
            for c in range(COLS):
                if grid[r][c] == 0:
                    q.append((r,c))
        directions = [[-1,0], [1,0],[0,1],[0,-1]]
        while q:
            length = len(q)
            
            for _ in range(length):
                r, c = q.popleft()

                for dr,dc in directions:
                    nr = dr + r
                    nc = dc + c
                    if nr in range(ROWS) and nc in range(COLS) and grid[nr][nc] > grid[r][c]:
                        grid[nr][nc] = grid[r][c] + 1
                        q.append((nr, nc))
